### PR TITLE
`defsetting` setter functions should end in `!`

### LINKED ISSUE #22799: `defsetting` setter functions should end in `!`
Right now if I do

```clj
(defsetting my-neat-setting)
```

if defines a function like

```clj
(defn my-setting ([] ...) ([new-value-or-nil] ...))
```

where the zero-arity fetches the current value and the one-arity updates the value

e.g.

```clj
(my-neat-setting) => nil
(my-neat-setting 100) ; set the value
(my-neat-setting) => 100
(my-neat-setting nil) ; unset the value
```
We should change the macro to generate two separate functions like `(defn my-setting [] ...)` and `(defn my-setting! [new-value-or-nil] ...)` instead e.g.

```clj
(my-neat-setting) => nil
(my-neat-setting! 100) ; set the value
(my-neat-setting) => 100
(my-neat-setting! nil) ; unset the value
```

I think that makes things more readable and follows the way we use exclamation points in the codebase. Honestly I'm not 100% sure why I didn't do it that way in the first place but this thought only occurred to me recently

### PR DESCRIPTION
**Pull request by camsaul**
Resolves #22799

`defsetting` setter functions now end in `!`; they are not defined if the Setting is defined with `:setter :none`. Clj-kondo will lint usage of the getter and setter functions and complain if they're passed the wrong number of args.

On top of this I improved our Kondo config a lot:

- Added a new custom `:analyze-call` hook for `defsetting` so Kondo knows about the getter and setter and can complain if they are called with the wrong number of args
- Added a new custom `:analyze-call` hook for `with-temporary-setting-values` so we get correct error messages in the correct places
- Told Kondo how to handle `mt/are+`
- Fixed the `:macroexpand` hook for `toucan.models/defmodel` which incorrectly mapped the table name and to a record field
- Fixed some of our custom macroexpansion namespaces not starting with `macros.` like the other ones did

I also did a bit of general unrelated code cleanup because I couldn't help myself:

- Removed unused `u/pdoseq` and `u/rpartial` macros
- Fixed tests failing silently because we were logging the error with `log/error` instead of `log/fatal` before calling `System/exit`
- Added extra tests to make sure `defsetting` and `define-multi-setting` preserve metadata (e.g. `^:private` or `^:deprecated`) for both the getter and setter functions
- Lots of small code cleanup/formatting fixes

### CODE REVIEW

File: .clj-kondo/hooks/metabase/models/setting.clj [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  (:require [clj-kondo.hooks-api :as hooks]))
+
+(defn defsetting
+  "Rewrite a [[metabase.models.defsetting]] form like
+
+    (defsetting my-setting \"Description\" :type :boolean)
+
+  as
+
+    (let [_args [\"Description\" :type :boolean]}
```
Comment (dpsutton) [inline .clj-kondo/hooks/metabase/models/setting.clj:11]: from reading the code, isn't this more like `(let [_ "Description" _ :type _ :boolean] (defn my-setting []) (defn my-setting! [_value-or-nil]))`

**Comment by dpsutton** [comment]
We currently have unused public var warnings. I asked and apparently we cannot emit the equivalent of `#_ {:clj-kondo/ignore [:unused-public-var]}` so we need to emit usages.

```clojure
(defmacro defmodel [model-name & _args]
  `(do
     (def ~model-name "docstring" nil)
     (defrecord ~(symbol (str model-name "Instance")) [])
     ~model-name
     (~(symbol (str "map->" model-name "Instance")) {})
     (~(symbol (str "->" model-name "Instance")))))
```

silences the warning on `defmodel`. I'm still working on how to do this in the hooks.

File: .clj-kondo/hooks/metabase/models/setting.clj (hunk 1 above)
Comment (camsaul) [reply .clj-kondo/hooks/metabase/models/setting.clj:11]: Oh yeah you're right.

Comment (camsaul) [inline .clj-kondo/hooks/metabase/models/setting.clj:11]:
```suggestion
    (let [_ \"Description\"
          _ :type 
          _ :boolean]
```

**Review by dpsutton — APPROVED** [review]

**Pull request merged**